File: Display Plugins/GrowlBezierPathAdditions.m
//
//  GrowlBezierPathAdditions.m
//  Display Plugins
//
//  Created by Ingmar Stein on 17.11.04.
//  Copyright 2004 __MyCompanyName__. All rights reserved.
//

#import "GrowlBezierPathAdditions.h"

@implementation NSBezierPath(GrowlBezierPathAdditions)
+ (NSBezierPath *)roundedRectPath:(NSRect)rect radius:(float)radius lineWidth:(float)lineWidth
{
	float inset = radius + lineWidth;
	NSRect irect = NSInsetRect( rect, inset, inset );
	float minX = NSMinX( irect );
	float minY = NSMinY( irect );
	float maxX = NSMaxX( irect );
	float maxY = NSMaxY( irect );

	NSBezierPath *path = [NSBezierPath bezierPath];
	[path setLineWidth:lineWidth];

	[path appendBezierPathWithArcWithCenter:NSMakePoint( minX, minY )
									 radius:radius 
								 startAngle:180.f
								   endAngle:270.f];

	[path appendBezierPathWithArcWithCenter:NSMakePoint( maxX, minY ) 
									 radius:radius 
								 startAngle:270.f
								   endAngle:360.f];

	[path appendBezierPathWithArcWithCenter:NSMakePoint( maxX, maxY )
									 radius:radius 
								 startAngle:0.f
								   endAngle:90.f];

	[path appendBezierPathWithArcWithCenter:NSMakePoint( minX, maxY )
									 radius:radius 
								 startAngle:90.f
								   endAngle:180.f];

	[path closePath];

	return( path );
}

@end

File: Core/Source/GrowlBezierPathAdditions.c
//
//  GrowlBezierPathAdditions.c
//  Display Plugins
//
//  Created by Ingmar Stein on 17.11.04.
//  Copyright 2004-2006 The Growl Project. All rights reserved.
//
// This file is under the BSD License, refer to License.txt for details

#include "GrowlBezierPathAdditions.h"

void addRoundedRectToPath(CGContextRef context, CGRect rect, CGFloat radius) {
	CGFloat minX = CGRectGetMinX(rect);
	CGFloat minY = CGRectGetMinY(rect);
	CGFloat maxX = CGRectGetMaxX(rect);
	CGFloat maxY = CGRectGetMaxY(rect);
	CGFloat midX = CGRectGetMidX(rect);
	CGFloat midY = CGRectGetMidY(rect);

	CGContextBeginPath(context);
	CGContextMoveToPoint(context, maxX, midY);
	CGContextAddArcToPoint(context, maxX, maxY, midX, maxY, radius);
	CGContextAddArcToPoint(context, minX, maxY, minX, midY, radius);
	CGContextAddArcToPoint(context, minX, minY, midX, minY, radius);
	CGContextAddArcToPoint(context, maxX, minY, maxX, midY, radius);
	CGContextClosePath(context);
}
